### ISSUE TITLE
Introduce a parameter initialization stage

### ISSUE THREAD

**Issue by ottonemo** [dnouri/skorch#342]
Since there is still discussion around how to properly and conveniently freeze parameters (#261, #337) I'd like to present a generic way of approaching this problem while avoiding callbacks (which may seem an unnatural choice for such an action, but that might also be personal preference) and opening up a number of modifications that were not easily possible.

The general idea is to introduce a parameter initialization stage that applies user-specified functions to certain parameters, e.g.:

```python
net = Net(
	MyModule,
	parameter_map={
		'linear[0-9].weight': skorch.helper.freeze_param,
		'linear[0-9].bias': skorch.helper.freeze_param,
		'rnn*.h2h': torch.init.xavier_normal_,
	}
	parameter_map__linear0__weight=torch.init.xavier_normal_,
)
```

this approach has the following benefits:

- works with both uninitialized and initialized modules
- compatible with grid search (`parameter_map__linear0__weight` notation)
- easily extensible to multiple optimizers (basically no change needed under the assumption of non-overlapping parameter sets between optimizers)
- allows easy test of different initialization schemes
- compatible with torch utility functions

while there are a few benefits it has the drawback of introducing a new specialized parameter that needs documentation and introduces a new workflow besides callbacks.

I'd like to hear your feedback.

Idea sketch:
```python
class _untrainable_marker: pass

def freeze_parameter(param):
	param.requires_grad = False
	return _untrainable_marker

def apply_parameter_map(self, map, named_parameters):
	for name, param in named_parameters:
		for map_name, map_fn in map.items():
			if not fnmatch.fnmatch(name, map_name):
				continue
			if map_fn(param) is _untrainable_marker:
				break
		else:
			continue
		yield name, param

class Net:
	def __init__(self, parameter_map=None, **kwargs):
		self.parameter_map = self._collect_parameter_map(parameter_map, kwargs)

	def collect_parameter_map(self, pmap, kwargs):
		for k, fn in self.get_params_for('parameter_map', kwargs).items():
			pmap[k.replace('__', '.')] = fn
		return pmap

	def initialize_parameters(self):
		self.trainable_parameters_ = apply_parameter_map(self.parameter_map, self.module_.named_parameters())

	def initialize_optimizer(self):
		self.optimizer_ = self.optimizer(self.trainable_parameters_)

	def initialize(self):
		self.init_callbacks()
		self.init_module()
		self.init_parameters()
		self.init_optimizer()

net = Net(
	MyModule,
	parameter_map={
		'linear[0-9].weight': skorch.helper.freeze_param,
		'linear[0-9].bias': skorch.helper.freeze_param,
		'rnn*.h2h': torch.init.xavier_normal_,
	}
	parameter_map__linear0__weight=torch.init.xavier_normal_,
)
```

**Comment by BenjaminBossan** [comment]
This approach would tick up the complexity but that is nothing that can't be handled by good naming and documentation. Moreover, it's a feature that many users probably don't need and if it doesn't get in their way, that's fine.

Some ideas need more flashing out -- I don't like how `_untrainable_marker` is used -- but it can certainly be overcome.

I would like stress an advantage of this approach in contrast of using callbacks for similar things: The earliest that a callback can have an effect is `on_train_begin`. If callbacks modify module parameters and this effect should be reflected in the optimizer (e.g. excluding frozen parameters), we would not to re-initialize the module and optimizer *after* `on_train_begin`. This is bad and I'm not even sure how that would work with `partial_fit`. Therefore, I think that callbacks cannot be part of a long term solution of the problem.

**Comment by ottonemo** [comment]
Good point.

# Callback approach redux

Since the optimizer now ignores values with `requires_grad=False` the callback route is an option again.
Ideally we would also want to support freezing specific layers at some point during the training.
While there is a use-case for having frozen layers at the beginning of training it is not the generic
use-case. Ideally we would have a mechanism that executes some function on some parameters at a given time.

For example:

    ParamFunc(
        'linear*.weight',
        at=lambda net: len(net.history) == 10,
        fn=skorch.helper.freeze_param,
    )

A simpler but less generic variant of this would be

    ParamFunc('linear*.weight', epoch=10, fn=skorch.helper.freeze_param)

Freezing parameters would then be as easy as

    ParamFunc('linear*.weight', epoch=0, fn=skorch.helper.freeze_param)

which could also be pre-defined

    Freezer = partial(ParamFunc, epoch=0, fn=skorch.helper.freeze_param)
    Initializer = partial(ParamFunc, epoch=0)

In a grid search this could look like this

    net = Net(mod, callbacks=[
        ('init_1', Initializer('linear*.weight', torch.nn.init.xavier_normal_)),
    ])

    params = {
        'callbacks__freezing': [
            Freezer(['linear*.weight', 'linear*.bias']),
            Freezer(['linear*.weight']),
            Freezer(['linear*.weight'], epoch=10),
            None,
        ],
        'callbacks__init_1__fn': [
            torch.nn.init.xavier_normal_,
            torch.nn.init.uniform_
        ],
        'callacks__init_2': [
            Initializer('linear*.weight', torch.nn.init.xavier_normal_),
            Initializer('linear*.weight', torch.nn.init.uniform_),
        ],
        'callbacks__reinit': [
            ParamFunc('linear*.weight', at=lambda net: net.history[-1, 'mse_loss'] < 1, fn=torch.nn.init.xavier_normal_),
            None,
        ],
    }

if one designs the interface of `ParamFunc` so that `at` has precedence over `epoch`
it would be possible to have something like

    Freezer('linear*.weight', at=lambda net: net.history[-1, 'mse_loss'] < 0.05)

for conditional freezing without resorting to the more generic `ParamFunc` (at the loss of semantics).

Another consideration is to support a function that returns the mutation function depending
on the state of the network, i.e. a mixture of `at` and `fn`:

    def sched(net):
        if 0 <= len(net.history) < 10:
            return skorch.util.freeze_param
        elif 10 <= len(net.history) < 20:
            return skorch.util.unfreeze_param
        else:
            return skorch.util.freeze_param

    cb = ParamFunc('linear*.weight', schedule=sched)

The order of precedence would be `epoch < at < schedule` since `epoch` is implemented by
`at` and `at` is implemented by `schedule`.

**Comment by benjamin-work** [comment]
> As a side note, the optimizers does not check for `require_grads` during initialization anymore: [pytorch/pytorch@f43e067](https://github.com/pytorch/pytorch/commit/f43e06712829ce24552d73e2442ef31385559729)

That is really surprising, as I remember a thread where the pytorch devs explicitely said they wouldn't do this. At least this worry is out of the way.

Overall, @ottonemo's proposal gets a 👍 from me. Although I still believe that setting attributes such as initialization of weights through a callback is a little bit counter-intuitive, the added benefit of switching things up at specific points in time is enough to convince me.

Regarding the `at` vs `epoch` vs `schedule`:
The former 2 could be solved by treating the parameter differently depending on whether it is an int or a callable. We have similar mechanisms at other places, so that shouldn't be too surprising.
With `schedule` I'm not sure because that seems to conflict with the `fn` parameter (what happens when both are specified?). Instead, I could see that `fn` must be a function that optionally takes `net` or `history` as second argument and can conditionally apply its modification depending on some property of the net/history.

Regarding the grid search example, it won't work as shown because it's impossible to set callbacks that were not specified during init (i.e. adding new callbacks). There would need to be placeholders with the given name, e.g. `callbacks=[('freezing', None), ...]`.

For `ParamFunc` we need a better name :)

One more concern: Are we certain that when passing all parameters to the optimizer and later on freezing some of them does not adversely affect memory usage? I would like to be certain of that before we add these changes.

**Comment by ottonemo** [comment]
>  The former 2 could be solved by treating the parameter differently depending on whether it is an int or a callable. We have similar mechanisms at other places, so that shouldn't be too surprising.

I like that the name is explicitly epoch which helps with semantics but I also see that it might clutter the parameter space.

> Instead, I could see that fn must be a function that optionally takes net or history as second argument and can conditionally apply its modification depending on some property of the net/history.

Nah, this would make using functions like `torch.nn.init.xavier_normal_` too difficult. I feel like `schedule` is the function that receives `at` and `fn` as parameters, so whenever the user supplies a schedule the user gets to decide what happens with these two parameters. The default implementation of `schedule` would then be something akin to

```python
def default_schedule(net, fn, at):
    if at(net):
        return fn
    return None
```

> Regarding the grid search example, it won't work as shown because it's impossible to set callbacks that were not specified during init (i.e. adding new callbacks). There would need to be placeholders with the given name, e.g. `callbacks=[('freezing', None), ...]`.

Is this a feature we want or is this to hard to do? I feel like it should be possible for the sake of convenience (say I want to quickly add a new schedule for a different parameter set then I only want to touch grid search params, not the net itself).

> For `ParamFunc` we need a better name :)

`ParamMapper` maybe since this callback is effectively doing `map(sched()(), filter(pattern, params))`?

**Comment by ottonemo** [comment]
Since #343 is merged now, we have a way to (hopefully) conveniently freeze/unfreeze parameters and do all sorts of other things along the way e.g. initialize parameters, periodically freeze/unfreeze parameters, re-initialize under some condition, normalize parameters if their norm is getting our of hand, etc.
